### PR TITLE
Signcontract regaddressfix merge

### PR DESCRIPTION
**Pull request by lawlawlaw**
Combine changes from the signcontract and regaddressfix branches.

The changes are:
From regaddressfix:
1. onboarding rpc tests are updated
2. fixes to the whitelisting (registeraddress) system
3. ECIES uses randomly generated IV instead of one generated from a hash of the ECDH shared secret. This guarantees a unique  IV even if an encryption key is reused. 
4. A unencrypted whitelisting system is introduced. Checks are included in init.cpp for illogical config option combinations.
From signcontract:
See the pull request

### CODE REVIEW

File: src/init.cpp [hunk 1]
```diff
@@ -145,7 +145,7 @@ bool ShutdownRequested()
 /**
  * This is a minimally invasive approach to shutdown on LevelDB read errors from the
  * chainstate, while keeping user interface out of the common library, which is shared
- * between bitcoind, and bitcoin-qt and non-server tools.
+ * between bitcoind, and bitcoin-qt and npkhwhiteliston-server tools.
```
Comment (nkostoulas) [inline src/init.cpp]: ?

File: src/init.cpp [hunk 2]
```diff
@@ -1446,6 +1464,10 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
     fReindex = GetBoolArg("-reindex", false);
     bool fReindexChainState = GetBoolArg("-reindex-chainstate", false);
 
+    if((fRequireWhitelistCheck || fScanWhitelist) &! (fReindex || fReindexChainState))
+        return InitError("-pkhwhitelist or -pkhwhitelist-scan require either -reindex-chainstate or -reindex");
```
Comment (nkostoulas) [inline src/init.cpp]: I thought we discussed reindex is not required for whitelisting.

Comment (nkostoulas) [reply src/init.cpp]: Has this changed?

File: src/policy/kycfile.h [hunk 3]
```diff
@@ -49,7 +49,7 @@ class CKYCFile{
    		 	ENCRYPTION_ERROR
   		};
 
-	 	bool getOnboardingScript(CScript& script);
+	 	bool getOnboardingScript(CScript& script, bool bBlacklist=false);
```
Comment (nkostoulas) [inline src/policy/kycfile.h]: convention is fBlacklist

File: src/policy/whitelist.cpp [hunk 4]
```diff
@@ -435,7 +286,8 @@ bool CWhiteList::RegisterDecryptedAddresses(const std::vector<unsigned char>& da
         if(!fEnd){
           CBitcoinAddress addrNew;
           std::vector<unsigned char> addrChars(itData1,itData2);
-          addrNew.Set(CKeyID(uint160(addrChars)));  
+          CTxDestination addr = CKeyID(uint160(addrChars));
+//          addrNew.Set(CKeyID(uint160(addrChars)));
```
Comment (nkostoulas) [inline src/policy/whitelist.cpp]: remove

File: src/policy/whitelist.cpp [hunk 5]
```diff
@@ -444,27 +296,29 @@ bool CWhiteList::RegisterDecryptedAddresses(const std::vector<unsigned char>& da
...
-            itData1=itData2;
-            if(!pubKeyNew.IsFullyValid())
-            {
-              itData1 = itStart;
-              itData2 = itStart;
-              if(pairsAdded == 0)
-                bEnd = true;
-              break;
-            }
+  //        std::string addrStr=addrNew.ToString();
```
Comment (nkostoulas) [inline src/policy/whitelist.cpp]: remove

Comment (lawlawlaw) [reply src/policy/whitelist.cpp]: Done

File: src/policy/whitelist.cpp (hunk 4 above)
Comment (lawlawlaw) [reply src/policy/whitelist.cpp]: Done

File: src/policy/kycfile.h (hunk 3 above)
Comment (lawlawlaw) [reply src/policy/kycfile.h]: Done

File: src/init.cpp (hunk 1 above)
Comment (lawlawlaw) [reply src/init.cpp]: Done

File: src/init.cpp (hunk 2 above)
Comment (lawlawlaw) [reply src/init.cpp]: I've corrected this for plain whitelist in the latest commit. Encrypted whitelist still requires reindex.

Comment (nkostoulas) [reply src/init.cpp]: Why?

Comment (lawlawlaw) [reply src/init.cpp]: Because I haven't changed CWhitelistEncrypted so its the same as before.

**Pull request merged**